src/Illuminate/Http/Client/Response.php

```diff
@@ -204,12 +204,27 @@ public function toPsrResponse()
      * Throw an exception if a server or client error occurred.
      *
      * @return $this
-     *
-     * @throws \Illuminate\Http\Client\RequestException
      */
     public function throw()
+    {
+        return $this->throwWith(function () {
+            //
+        });
+    }
+
+    /**
+     * Execute the callback and throw an exception if a server of client error occurred.
+     *
+     * @param \Closure $callback
+     * @return $this
+     *
+     * @throw \Illuminate\Http\Client\RequestException
+     */
+    public function throwWith($callback)
```

Comment:
What about `beforeThrowing`?

Reply:
Then it'd make sense to call `beforeThrowing()->throw()`, which IMO is too bervose

Reply:
agreed on that being to verbose. I was hoping to find something that could express "When throwing, do this thing and then throw" rather than two methods.

Reply:
`onError` also makes sense to me. `throwWith` doesn't make sense to me because it has an indication of which parameters would be given to `throw` (like the Cotnainer's `makeWith`). Though I still prefer `beforeThrowing` as it indicates what to do (callback) right before an exception is thrown. I guess `whenThrowing` could also work (like Query Builder `when()`).

Reply:
I don't disagree with your objections. As mentioned I didn't know if this was really a great method name.

Maybe there just isn't a nice "all-in-one" option, as `whenThrowing` also makes me wonder if I need to call `throw` or not.

Maybe it is just a `onFailure` callback (as others suggested on Twitter).

```php
return $client->withHeaders($headers)
    ->post($url, $payload)
    ->onFailure(fn ($response) =>
        Log::error('Twitter API failed posting Tweet', [
            'url' => $url,
            'payload' => $payload,
            'headers' => $headers,
            'response' => $response->body(),
        ])
    )->throw()->json();
```

Or maybe I just need to wrap up the more verbose middleware as @sebdesign pointed out in my own app 🤷‍♂️

I just liked how the failure and throw was on the other side of the `post` call, as it  _felt_ nicer to me, but that is just an opinion that might not be shared.

Reply:
Personally I prefer the `onError` callback instead of the `Middleware::mapResponse` stuff. It feels nicer and you get to work with the `Laravel\Http\Client\Response` instead of the Guzzle response.